### PR TITLE
Give existing deprecations an expiration date

### PR DESCRIPTION
**Pull request by WardBrian**
This is the first of several PRs to begin the implementation of https://github.com/stan-dev/design-docs/pull/43

This PR simply adds notice to all existing minor deprecations that they will be removed in Stan 2.32.0. This does not apply to the ODE integrators, only the syntax-level changes which can all be replaced with the canonicalizer.

Follow on PRs will:

- Add warnings for 'future keywords': at the very least `array` (#953), but also `offset`, `multiplier`, etc (basically re-doing https://github.com/stan-dev/stan/issues/2712)
- Mark old array syntax as deprecated. This is already stated in the docs but currently produces no warnings. 
  As part of this, I think we should run _at least_ the auto-formatter on our test models, if not also the canonicalizer, to prevent the test output from ballooning due to all the old models.

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [x] If a user-facing facing change was made, the documentation PR is here: There is one docs PR for all 3 of these stanc3 PRs: https://github.com/stan-dev/docs/pull/439
 
## Release notes

Marked existing syntax deprecations as expiring in Stan 2.32.0 in accordance with https://github.com/stan-dev/design-docs/pull/43. Users should update their models (manually or with `stanc --print-canonical`) _before_ that version of Stan.

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: src/frontend/Deprecation_analysis.ml [hunk 1]
```diff
@@ -85,50 +88,52 @@ let rec collect_deprecated_expr deprecated_userdefined
...
+             2.32.0. Use the conditional operator (x ? y : z) instead; this \
+             can be automatically changed using stanc --print-canonical" ) ]
       @ List.concat
           (List.map l ~f:(fun e ->
                collect_deprecated_expr deprecated_userdefined [] e ) )
   | FunApp (StanLib _, {name; _}, l) ->
       let w =
         if Option.is_some (String.Map.find deprecated_distributions name) then
+          let rename, version =
+            String.Map.find_exn deprecated_distributions name in
```
Comment (nhuurre) [inline src/frontend/Deprecation_analysis.ml:112]: Instead of doing `Map.find` twice you should just `match .. with` the first result. Also I think `Map.find` is preferred over `String.Map.find`.

File: src/frontend/Deprecation_analysis.ml [hunk 2]
```diff
@@ -141,15 +146,18 @@ let rec collect_deprecated_expr deprecated_userdefined
         let type_ = String.Map.find deprecated_userdefined name in
         if Option.is_some type_ then
           [ ( emeta.loc
-            , "Use of the _log suffix in user defined function " ^ name
-              ^ " is deprecated, use "
+            , "Use of the _log suffix in user defined probability function "
+              ^ name
+              ^ " is deprecated and will be removed in Stan 2.32.0, use name '"
               ^ update_suffix name (Option.value_exn type_)
```
Comment (nhuurre) [inline src/frontend/Deprecation_analysis.ml:152]: Same here: don't do `if Option.is_some type_ then Option.value_exn type_` but instead do `match type_ with Some ..`

File: src/frontend/Deprecation_analysis.ml [hunk 3]
```diff
@@ -85,50 +88,52 @@ let rec collect_deprecated_expr deprecated_userdefined
...
-          , "The no-argument function `get_lp()` is deprecated. Use the \
-             no-argument function `target()` instead." ) ]
   | FunApp (StanLib FnPlain, {name= "abs"; _}, [e])
     when Middle.UnsizedType.is_real_type e.emeta.type_ ->
       collect_deprecated_expr deprecated_userdefined
         ( acc
         @ [ ( emeta.loc
             , "Use of the `abs` function with real-valued arguments is \
-               deprecated; use functions `fabs` instead." ) ] )
+               deprecated; use function `fabs` instead." ) ] )
```
Comment (nhuurre) [inline src/frontend/Deprecation_analysis.ml:97]: Isn't `fabs()` going to be deprecated soon? We won't need this `abs()` branch at all.

File: src/frontend/Deprecation_analysis.ml [hunk 4]
```diff
@@ -176,6 +184,20 @@ let rec collect_deprecated_stmt deprecated_userdefined
       acc
       @ collect_deprecated_lval deprecated_userdefined [] l
       @ collect_deprecated_expr deprecated_userdefined [] e
+  | Tilde {distribution= {name; id_loc}; arg; args; _}
+    when Option.is_some (String.Map.find deprecated_userdefined (name ^ "_log"))
+    ->
```
Comment (nhuurre) [inline src/frontend/Deprecation_analysis.ml:189]: One more `Option.(if is_some .. then value_exn ..)`. I also don't understand why this is here. It looks like this is detecting if the tilde statement refers to a user-defined `_log` density function. But the `_log` suffix is not visible here; in my opinion it would make more sense to warn about the function definition, not the places that use the function.

File: src/frontend/Warnings.ml [hunk 5]
```diff
@@ -7,8 +7,8 @@ let pp ?printed_filename ppf (span, message) =
   let loc_str =
     if span = Location_span.empty then ""
     else " in " ^ Location.to_string ?printed_filename span.begin_loc in
-  Fmt.pf ppf "@[<hov>Warning%s: %s@]" loc_str message
+  Fmt.pf ppf "@[<hov 4>Warning%s: %a@]" loc_str Fmt.text message
 
 let pp_warnings ?printed_filename ppf warnings =
   if List.length warnings > 0 then
-    Fmt.(pf ppf "@[<v>%a@]@." (list ~sep:cut (pp ?printed_filename)) warnings)
+    Fmt.(pf ppf "@[<v>%a@]@\n" (list ~sep:cut (pp ?printed_filename)) warnings)
```
Comment (nhuurre) [inline src/frontend/Warnings.ml:14]: Why this change? We try to avoid `\n` in format strings because the formatter does not recognize it as a newline.

File: src/frontend/Deprecation_analysis.ml (hunk 1 above)
Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml:112]: Yeah, the structure of this entire file is really awful for this reason. I was trying to avoid redoing all of it, especially because it will now mostly be deleted a couple versions from now, but it's probably worth it.

File: src/frontend/Deprecation_analysis.ml (hunk 3 above)
Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml:97]: I'm not sure about the status of #475 so I just left this unchanged for now. It should be handled alongside the changes to the supported signatures

File: src/frontend/Deprecation_analysis.ml (hunk 4 above)
Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml:189]: We currently warn if a user writes 

`target += foo_log(x, ...)`
but not
`x ~ foo(...)` 

even though this are completely equivalent. Really it's the second one which is actually being deprecated/removed - if a user still wants to name a function `bar_log` in the future they can, it's just that it won't have the special meaning it does now. Warning at the definition time would suggest that they won't be able to do such a thing

File: src/frontend/Warnings.ml (hunk 5 above)
Comment (WardBrian) [reply src/frontend/Warnings.ml:14]: `@\n` and `\n` are not the same, the former works like `@.` but always forces a newline rather than letting the formatter sometimes ignore it. In particular, when ppf is stderr, this guarantees that the formatter always flushes the stream.

File: src/frontend/Deprecation_analysis.ml (hunk 4 above)
Comment (nhuurre) [reply src/frontend/Deprecation_analysis.ml:189]: The warning at the definition suggest the meaning of the definition is going to change. Currently `_log` functions can use `_lupdf` and that's going to be deprecated. (which btw means that `target += foo_log()` is not _completely_ equivalent to a tilde)
I also think that use in tilde statement indicates a problem with the function definition which no longer matches the intended use.

File: src/frontend/Deprecation_analysis.ml (hunk 1 above)
Comment (nhuurre) [reply src/frontend/Deprecation_analysis.ml:112]: If you decide rewriting all this is not worth it then I'd still change that `Option.is_some  (String.Map.find ..)` to `Map.mem`

File: src/frontend/Deprecation_analysis.ml (hunk 4 above)
Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml:189]: So are y0u suggesting we not warn at any usages, only at the definitions?

Comment (nhuurre) [reply src/frontend/Deprecation_analysis.ml:189]: I think
* Don't warn at calls. Those warnings are redundant if the definition emits a warning and unnecessary if the definition does not need to change.
* Don't warn at tildes. The tilde does not mention the deprecated suffix so even if the user takes action the statement itself will not change.
* The definition must warn if it either contains `_lupdf` or is used in a tilde, otherwise it's fine as is. That's kind of tricky to check, and the canonicalizer updates the name unconditionally so "always warn" is a reasonable choice here.

Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml]: I guess the tricky thing is that `_log` isn't technically a reserved suffix as far as I am aware, so the user should be able to write something like

```stan
functions {
   real add_one_log(real x){
     return log1p(x);
   }
}

model {
   real x = add_one_log(3.);
}
```

without it warning anywhere (though it looks like syntax highlighters are too broad on this as well), as this will be the sort of thing _log could totally be used for now and in the future. (as a historical note, it's not clear to me why both the _log suffix and the function multiply_log _both_ needed to be deprecated, but here we are)

If we want, I can try to make it warn only for Tildes and in the body of a definition where lup[dm]f is called? I agree that isn't the ideal location, but it should at least be doable and scoped how we want it to be.

Comment (nhuurre) [reply src/frontend/Deprecation_analysis.ml]: Yes, that's the problem. I'd rather just complain during the transition period and then allow it after `_log` distributions are completely gone. While your proposal is doable for the warnings module I don't see how to make it work in the canonicalizer.

Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml]: I've settled on warning at the definition, _only_ when:

1. The name ends in "_log"
2. The return type is either UReal or UInt

These are necessary conditions for it to be in the deprecation case, but sadly not sufficient to say it isn't a false alarm. I've updated the warning to account for this:

> ```
> Warning in 'log.stan', line 2, column 8: Use of the _log suffix in user
>     defined probability functions is deprecated and will be removed in Stan
>     2.32.0, use name 'add_one_log_lpdf' instead if you intend on using this
>     function in ~ statements or calling unnormalized probability functions
>     inside of it.
> ```

Canonicalize behavior also updated to make sure the return type is either real or int, otherwise it just leaves the name. Does this sound good?

Comment (nhuurre) [reply src/frontend/Deprecation_analysis.ml]: A function with `_log` suffix must have UReal return type, otherwise it does not pass typecheck
https://github.com/stan-dev/stanc3/blob/9802b07906eaa2014c59a1fdf7e5aa6747f247b4/src/frontend/Typechecker.ml#L1257-L1265

Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml]: Oh, what's all this business then? It seems like it's just wrong?

https://github.com/stan-dev/stanc3/blob/f1c30d28e143f4fe3fe790eb38c5bda872306c6a/src/frontend/Deprecation_analysis.ml#L75-L77

Comment (nhuurre) [reply src/frontend/Deprecation_analysis.ml]: That should not be the return type but the type of the first argument. Also I'm not sure if `is_real_type` is exactly correct there either, depends on if deeply nested arrays are allowed or not...

Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml]:
```ocaml
let is_real_type = function
  | UReal | UVector | URowVector | UMatrix
   |UArray UReal
   |UArray UVector
   |UArray URowVector
   |UArray UMatrix ->
      true
  | _ -> false
```

besides my mixup, does that sound like an okay way to proceed?

Comment (nhuurre) [reply src/frontend/Deprecation_analysis.ml]: Oh, wow.
```stan
real foo_log(array[,] real x) {
  return sum(x[1]);
}
real foo_lpdf(array[,] real x) {
  return sum(x[1]);
}
```
The first one is fine, the second is apparently not? This is surely a bug but which one is wrong?

Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml]: Well that's certainly not ideal. I just checked and the same behavior is in 2.27, so it's not even something introduced by #995, it's just been around a while.

Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml]: I think it's because _log could be either a lpmf or lpdf, so it would be ambiguous whether it should be a real or int, so nothing is checked at all. I can trivially fix it in a separate PR to check that it's at least an int or a real but nothing else.

Comment (nhuurre) [reply src/frontend/Deprecation_analysis.ml]: I'm not sure if we want to be so restrictive. What's wrong with distribution of an array of arrays? `UnsizedType.is_real_type` and `is_int_type` could be changed to recurse deep into array types, that's how I'd expect them to behave based on their names anyway.

Comment (WardBrian) [reply src/frontend/Deprecation_analysis.ml]: I agree, as does @bob-carpenter. I've opened #1045 about it

**Comment by WardBrian** [comment]
Bugs in the typechecker aside, I think this is good to go @nhuurre

**Review by nhuurre — APPROVED** [review]
